### ISSUE TITLE
swift url double slash

### ISSUE THREAD

**Issue by carraher** [wal-e/wal-e#308]
swift protocol shows a good URL in log output
```
  DETAIL: Uploading "pg_xlog/000000010000000000000001" to "swift://deis-stg-database/wal_005/000000010000000000000001.lzo".
```

Later this is transformed into http(s) protocol along with the FQDN of the Swift endpoint and with an extra '/' between the container and object path.
This works okay with some swift providers (OpenStack works). However others it does not.
```
swiftclient.exceptions.ClientException: Object PUT failed: https://my.uri.com:8083/v1/tenant/deis-stg-database//wal_005/000000010000000000000001.lzo 412 Precondition Failed  [first 60 chars of response] b'<html><h1>PreconditionFailed</h1><p>At least one of the prec'
```

I'm not sure if Swift should work this way, but the extraneous slash is not likely intended.
As stated OpenStack does allow it even if incorrect
```
curl -k -X PUT -H "X-Auth-Token: $X_AUTH_TOKEN" "$X_STORAGE_URL/test-container//test-object-2" -d 'hello'
curl -k -H "X-Auth-Token: $X_AUTH_TOKEN" "$X_STORAGE_URL/test-container"
    /test-object-2
    test-object
curl -k -H "X-Auth-Token: $X_AUTH_TOKEN" "$X_STORAGE_URL/test-container//test-object-2"
    hello
```

**Comment by fdr** [comment]
You are right, that is not intentional. Care to submit a patch?

**Comment by listingmirror** [comment]
I believe GCE has an extra / also. Makes an empty / folder. Changing this would be a massively breaking change though.,

**Comment by fdr** [comment]
Yeah. It would break old backups and pruning. I wonder if it could be made backwards compatible via environment variable and then be deprecated over two versions. Most backups are aged out in a "reasonable" amount of time, so I don't think the matter is unfixable.

**Comment by carraher** [comment]
Is there any chance this is "safe" to fix in an upcoming release? The only other choice is for object storage providers to replicate OpenStack Swift's un-spec'd functionality.
For restoring legacy backups, can the restore process squash all double-slash `//` to single `/` ? Or do we not have programmatic hooks to get in the middle of restore?

**Comment by fdr** [comment]
On Tue, Jun 13, 2017 at 9:03 AM carraher <notifications@github.com> wrote:

> Is there any chance this is "safe" to fix in an upcoming release? The only
> other choice is for object storage providers to replicate OpenStack Swift's
> un-spec'd functionality.
>
I mostly need someone willing to test changes thoroughly. If we can write a
changelog for the the next release saying to, say, write the swift URL with
an extra slash, I'd consider that acceptable.

>
